scripts/lib/upgrade-zulip-stage-2

```diff
@@ -280,3 +280,7 @@ if not args.skip_purge_old_deployments:
     subprocess.check_call(["./scripts/purge-old-deployments"])
 else:
     logging.info("Skipping purging old deployments.")
+
+if args.skip_puppet:
+    logging.info("Showing un-applied puppet changes:")
+    subprocess.check_call(["./scripts/zulip-puppet-apply", "--force", "--noop", "--show_diff"])
```

Comment:
This feels like a weird way to express this idea, with `--force` and friends; maybe we want `zulip-puppet-apply` to expose a `--noop` option that means these 3 things?  `--force --noop` sounds kinda scary.

Reply:
The problem is that `--force` is really `--no-artificially-imposed-extra-verification`, which is what makes it sound scary.  I'm a little leery of having `--noop` imply `--force`, since then we start down the road of trying to duplicate the `puppet apply` option parsing -- i.e. does it have a short form?  Do we correctly support passing unambiguous but truncated option names?

...actually, doing some testing, `--show` _nominally_ works for `--show-diff`.  In actuality, however, it accepts the option as long as it is unambugious, but *only does the right thing if it's spelled out in full*.  That is:
```
root@alexmv-prod:/home/zulip/deployments/current# puppet apply --modulepath /home/zulip/deployments/current/puppet -e 'Exec { path => "/usr/sbin:/usr/bin:/sbin:/bin" } include zulip::voyager' --noop \
  --s
Error: Could not parse application options: ambiguous option: --s
```
```
root@alexmv-prod:/home/zulip/deployments/current# puppet apply --modulepath /home/zulip/deployments/current/puppet -e 'Exec { path => "/usr/sbin:/usr/bin:/sbin:/bin" } include zulip::voyager' --noop \
  --show
Notice: Compiled catalog for alexmv-prod.zulipdev.org in environment production in 1.42 seconds
Notice: /Stage[main]/Zulip::Apt_repository/Exec[setup_apt_repo]/returns: current_value 'notrun', should be ['0'] (noop)
Notice: Class[Zulip::Apt_repository]: Would have triggered 'refresh' from 1 event
Notice: /Stage[main]/Zulip::App_frontend/File[/etc/nginx/sites-available/zulip-enterprise]/content: current_value '{md5}292138adea408171b64e29fa0141a84b', should be '{md5}062d03287b49273f099a14ea88cc3339' (noop)
Notice: Class[Zulip::App_frontend]: Would have triggered 'refresh' from 1 event
Notice: /Stage[main]/Zulip::Nginx/Service[nginx]: Would have triggered 'refresh' from 1 event
Notice: Class[Zulip::Nginx]: Would have triggered 'refresh' from 1 event
Notice: Stage[main]: Would have triggered 'refresh' from 3 events
Notice: Applied catalog in 0.53 seconds
```
```
root@alexmv-prod:/home/zulip/deployments/current# puppet apply --modulepath /home/zulip/deployments/current/puppet -e 'Exec { path => "/usr/sbin:/usr/bin:/sbin:/bin" } include zulip::voyager' --noop \
  --show_diff
Notice: Compiled catalog for alexmv-prod.zulipdev.org in environment production in 1.48 seconds
Notice: /Stage[main]/Zulip::Apt_repository/Exec[setup_apt_repo]/returns: current_value 'notrun', should be ['0'] (noop)
Notice: Class[Zulip::Apt_repository]: Would have triggered 'refresh' from 1 event
Notice: /Stage[main]/Zulip::App_frontend/File[/etc/nginx/sites-available/zulip-enterprise]/content:
--- /etc/nginx/sites-available/zulip-enterprise	2020-08-01 06:36:23.400938982 +0000
+++ /tmp/puppet-file20200801-14346-mqx2ib	2020-08-01 06:38:18.890397154 +0000
@@ -34,4 +34,3 @@
     include /etc/nginx/zulip-include/certbot;
     include /etc/nginx/zulip-include/app;
 }
-

Notice: /Stage[main]/Zulip::App_frontend/File[/etc/nginx/sites-available/zulip-enterprise]/content: current_value '{md5}292138adea408171b64e29fa0141a84b', should be '{md5}062d03287b49273f099a14ea88cc3339' (noop)
Notice: Class[Zulip::App_frontend]: Would have triggered 'refresh' from 1 event
Notice: /Stage[main]/Zulip::Nginx/Service[nginx]: Would have triggered 'refresh' from 1 event
Notice: Class[Zulip::Nginx]: Would have triggered 'refresh' from 1 event
Notice: Stage[main]: Would have triggered 'refresh' from 3 events
Notice: Applied catalog in 0.44 seconds
```

So maybe if `puppet apply` itself doesn't get this right, we don't need to worry.  I guess I'm fine to add `--noop` as a first-class term.  We also already use `--force --noop` in `scripts/lib/install`, so I'll adjust that.

Reply:
Yeah, I think that makes sense.  `puppet` unfortunately somehow has a lot of rough edges in the command line; I think the reason is basically that puppet-the-company only cares about the "Run it on lots of systems via a web UI" product they've built on top of it, so the CLI very stale.